File: BasicCalculatorII.cpp
#include<iostream>
#include<string>
#include<vector>

using namespace std;

class Solution {
public:
    int calculate(string s) {
        
        vector<int> nums;
        vector<char> operators;
        
        for(int i=0;i<s.length();i++){
            char c = s[i];
            if(c == ' ') continue;
            if(c >= '0' && c <= '9'){
                int num = 0;
                while(i < s.length() && s[i] >= '0' && s[i] <= '9'){
                    num = num * 10 + s[i] - '0';
                    i ++;
                }
                i --;
                if(!operators.empty() && (operators.back() == '*' || operators.back() == '/')){
                    int pre = nums.back();
                    nums.pop_back();
                    if(operators.back() == '*') num = pre * num;
                    else num = pre / num;
                    operators.pop_back();
                }
                nums.push_back(num);
            }else{
                operators.push_back(c);
            }
        }
        
        int result = nums.front();
        for(int i=0;i<operators.size();i++){
            if(operators[i] == '+'){
                result += nums[i+1];
            }else{
                result -= nums[i+1];
            }
        }
        
        return result;
    }
};

int main(){
    
    string s;
    getline(cin, s);
    
    Solution *solution = new Solution();
    cout<<solution->calculate(s);
    
    return 0;
    
}

File: BasicCalculatorII.py
class Solution:
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack = []
        
        i = 0
        sign = '+'
        while i < len(s):
            if s[i].isdigit():
                j = i + 1
                while j < len(s) and s[j].isdigit():
                    j += 1
                num = int(s[i : j])
                i = j - 1
                if sign == '+':
                    stack.append(num)
                elif sign == '-':
                    stack.append(-num)
                elif sign == '*':
                    stack[-1] = stack[-1] * num
                elif sign == '/':
                    if stack[-1] > 0:
                        stack[-1] = stack[-1] // num
                    else:
                        stack[-1] = -((-stack[-1]) // num)
            elif s[i] != ' ':
                sign = s[i]
            print(stack)
            i += 1
        return sum(stack)
        
if __name__ == '__main__':
    s = raw_input()
    solution = Solution()
    print(solution.calculate(s))
